core/src/main/java/org/apache/iceberg/Snapshots.java

```diff
@@ -0,0 +1,226 @@
+      this.startFileIndex = startFileIndex;
+      this.endFileIndex = endFileIndex;
+      this.sizeInBytes = sizeInBytes;
+      this.tasks = tasks;
+      this.lastIndexOfSnapshot = lastIndexOfSnapshot;
+    }
+
+    public long snapshotId() {
+      return snapshotId;
+    }
+
+    public int startFileIndex() {
+      return startFileIndex;
+    }
+
+    public int endFileIndex() {
+      return endFileIndex;
+    }
+
+    public long sizeInBytes() {
+      return sizeInBytes;
+    }
+
+    public CloseableIterable<FileScanTask> tasks() {
+      return tasks;
+    }
+
+    public boolean lastIndexOfSnapshot() {
+      return lastIndexOfSnapshot;
+    }
+  }
+
+  public static class MicroBatchBuilder {
+    private final Snapshot snapshot;
+    private final FileIO io;
+    private Expression rowFilter;
+    private boolean caseSensitive;
+    private Map<Integer, PartitionSpec> specsById;
+
+    private MicroBatchBuilder(Snapshot snapshot, FileIO io) {
+      this.snapshot = snapshot;
+      this.io = io;
+      this.rowFilter = Expressions.alwaysTrue();
+      this.caseSensitive = true;
+    }
+
+    public MicroBatchBuilder caseSensitive(boolean sensitive) {
+      this.caseSensitive = sensitive;
+      return this;
+    }
+
+    public MicroBatchBuilder filter(Expression newRowFilter) {
+      this.rowFilter = newRowFilter;
+      return this;
+    }
+
+    public MicroBatchBuilder specsById(Map<Integer, PartitionSpec> specs) {
+      this.specsById = specs;
+      return this;
+    }
+
+    public MicroBatch generate(int startFileIndex, long targetSizeInBytes, boolean isStarting) {
+      List<ManifestFile> manifests = isStarting ? snapshot.dataManifests() :
+          snapshot.dataManifests().stream().filter(m -> m.snapshotId().equals(snapshot.snapshotId()))
+              .collect(Collectors.toList());
+
+      List<Pair<ManifestFile, Integer>> manifestIndexes = indexManifests(manifests);
+      List<Pair<ManifestFile, Integer>> skippedManifestIndexes = skipManifests(manifestIndexes, startFileIndex);
+
+      return generateMicroBatch(skippedManifestIndexes, startFileIndex, targetSizeInBytes, isStarting);
+    }
+
+    private List<Pair<ManifestFile, Integer>> indexManifests(List<ManifestFile> manifestFiles) {
+      int currentFileIndex = 0;
+      List<Pair<ManifestFile, Integer>> manifestIndexes = Lists.newArrayList();
+
+      for (ManifestFile manifest : manifestFiles) {
+        int filesCount = manifest.addedFilesCount() + manifest.existingFilesCount();
+        manifestIndexes.add(Pair.of(manifest, currentFileIndex));
+        currentFileIndex += filesCount;
+      }
+
+      return manifestIndexes;
+    }
+
+    private List<Pair<ManifestFile, Integer>> skipManifests(List<Pair<ManifestFile, Integer>> indexedManifests,
+                                                            int startFileIndex) {
+      if (startFileIndex == 0) {
+        return indexedManifests;
+      }
+
+      int index = 0;
```

Comment:
Minor: it would improve readability to use `manifestIndex` to distinguish from data file indexes.

Reply:
Changed.